Makefile: Add install and install-user targets

amboar opened the pull request:
Script some of the behaviour previously captured in the readme as
instructions to execute manually.

Signed-off-by: Andrew Jeffery <andrew@aj.id.au>

amboar committed:
Makefile: Add install-user target

Script some of the behaviour previously captured in the readme as
instructions to execute manually.

Signed-off-by: Andrew Jeffery <andrew@aj.id.au>

hedayat commented on Makefile:
```diff
@@ -93,3 +93,12 @@ test-docs:
 	jshint --config .jshint.cfg extension/
+
+.PHONY: install
+install: dist
```
I think install target should depend on `compile`, and copy files directly to target directory. No need to create and archive and then extract it again!

benjaoming replied:
This works fine though? And `dist` already exists? `dist` might do other things in the future, so I think it's best to keep stuff clean: `build=>dist=>install`, that's AFAIK a normal procedure.

hedayat replied:
Well, `compile` already exists too. And no, `build=>install` is a normal procedure. `dist` target is usually used to create distribution packages (e.g. in autotools world), and is usually not used for anything else. Actually, I've never seen `build=>dist=>install` myself. 

Anyway, I gave my suggestion, and I won't push for it much. But still, I think that makes more sense.

hedayat commented on Makefile:
```diff
@@ -93,3 +93,12 @@ test-docs:
+	# Remove any old installation
+	rm -rf ${HOME}/.local/share/gnome-shell/extensions/contact@projecthamster.org
+	# Create directory
+	mkdir -p ${HOME}/.local/share/gnome-shell/extensions/contact@projecthamster.org
```
I think it should be installed to system-wide directory by default. You might make it possible to specify if it should be installed system-wide or into user's home directory.

Also, it should be possible to provide an installation prefix (DESTDIR) at least for system-wide installations. 

Thanks!

benjaoming replied:
That's not the point of this PR: It's automating the instructions for a local user installation.

@amboar will change the name of the make target to `install-user`. The other name is thus free for implementing a system-wide installation for some other purpose, perhaps in a fakeroot for packaging.

hedayat replied:
Well, the title and description doesn't tell it, at least explicitly. And, it is fairly easy to add. But if  the author is not willing to provide that, I'm fine.`

amboar replied:
> Well, the title and description doesn't tell it, at least explicitly

That's because the request to switch to `install-user` came after I made the PR. I've agreed to make the change.

benjaoming commented on Makefile:
```diff
@@ -93,3 +93,12 @@ test-docs:
 	jshint --config .jshint.cfg extension/
+
+.PHONY: install
+install: dist
```
Great! I wrote the steps in the README, happy to see them in the Makefile! Can we rename this to `install-user` ? It's because `sudo make install` won't do a system-wide installation which might be expected (that needs a slightly different path, but it's also doable).

amboar replied:
Right, I initially called it `install-local` but I think your suggestion of `install-user` is better. I'll rework it shortly.

benjaoming reviewed (changes requested):
Love this! Thanks @amboar! Wondering if `make install` is better as `make install-user` or similar ?

mwilck commented on Makefile:
```diff
@@ -93,3 +93,12 @@ test-docs:
+.PHONY: install
+install: dist
+	# Remove any old installation
+	rm -rf ${HOME}/.local/share/gnome-shell/extensions/contact@projecthamster.org
```
I'd rather do this in a separate step, call it `clean-install`, `uninstall-local`, or whatever. Deleting stuff in the user's home directory, outside the build dir itself, might come to an unpleasant surprise to some people.

benjaoming replied:
I think it's expected from `make install-user` to delete the existing extension.

> Deleting stuff in the user's home directory, outside the build dir itself, might come to an unpleasant surprise to some people.

This isn't "stuff", this is `${HOME}/.local/share/gnome-shell/extensions/contact@projecthamster.org`

Someone running `make install-user` shouldn't be surprised that this command overwrites an existing extension. That's what all software installers would do: Upgrade an existing previous installation.

mwilck commented on Makefile:
```diff
@@ -93,3 +93,12 @@ test-docs:
+	rm -rf ${HOME}/.local/share/gnome-shell/extensions/contact@projecthamster.org
+	# Create directory
+	mkdir -p ${HOME}/.local/share/gnome-shell/extensions/contact@projecthamster.org
+	# Unpack build
```
I'd prefer unpacking the zip inside some subdir of the build directory itself, and just using a symlink from `${HOME}/.local/share/gnome-shell/extensions` to that subdir. (This is how I do it here :grin:).

Whether `make` should create that symlink should or just print a message informing the user how to do it would be up to discussion. I'd prefer the latter.

(Edit) actually, the sources are already unpacked in the final form under `build`. So creating symlink to that directory would be sufficient. Not sure if we want that though, `build` might be overwritten with experimental code.

benjaoming replied:
@mwilck I don't think a normal user wants that?

I expect to be able to delete the whole build structure after installing.

mwilck reviewed (changes requested):
See inline comments.

hedayat commented:
Oops, sorry guys. I wrote my comments 4days ago, but didn't know that I should press "Submit review" :D

DirkHoffmann commented:
Trying to summarise what was written above: There are three targets,

- `install-user` : to `~/.local/share/gnome-shell/extensions/contact@projecthamster.org/`
- `install` : system-wide
- `dist` : create ZIP for submission to extensions.gnome.org

and they should all depend on `compile`. And `install*` depends on `dist`.

Sounds plausible to me.

hedayat commented:
>     * `dist` : create ZIP for submission to extensions.gnome.org
> 
> and they should all depend on `compile`. And `install*` depends on `dist`.

You already contradicted yourself: install* depends on dist, while dist creates an archive for submission to EGO (and install* doesn't submit the archive to EGO) :D ;)

Anyway, if you all like that, go with it. Simply, I don't. :)

benjaoming commented:
@mwilck 

> Before we merge this, we should clarify if we want to have it on develop only or also on our other branches. In the latter case, we need to create the branches first, and start merging to the "lowest" one we want to support.

It's true, removing the old instructions from README and replacing them with instructions that only work with the current `develop` branch isn't ideal.

@amboar What do you think about keeping the old instructions and instead add something like this?

> #### Shortcut on `develop`
>
> If you are using the `develop` branch since May 2020, you can run `make install-user` to install your current working branch in your user environment

amboar committed:
Makefile: Add install target

`make install` deploys the extension to the system path rather than
confining it to per-user directories.

Signed-off-by: Andrew Jeffery <andrew@aj.id.au>

amboar commented:
To be honest I didn't expect the PR to drive so much discussion. The patch made my life easier testing @mwilck 's 3.36 changes and I figured I might as well propose it. I'm not attached to it and am happy to drop it if we feel it's taking more energy to handle than is necessary.

I feel like applying it to `develop` only causes a bit of a maintenance headache, though I'm only guessing at how you manage your branches. That said I'm not living with the maintenance problem so I don't mind if you want to pick it up. I'll rework the readme as you suggest @benjaoming.

Addressing the `dist` vs `compile` target discussion, reusing `dist` was the path of least resistance as that was what was described in the readme. I think the question of whether we should conform more accurately to autotools semantics can be sorted out separately? I agree that how `dist` is implemented here doesn't really align with the `dist` target of autotools.

amboar committed:
README: Add description of install targets

The Makefile has been updated with ``install`` and ``install-user``
targets. Mention them in the README to save those tinkering with the
``develop`` branch from manually running the instructions.

Signed-off-by: Andrew Jeffery <andrew@aj.id.au>

amboar commented:
So with what I've just pushed we now have both `install` and `install-user` targets. I've split the patches out so you can take as many as you think are reasonable :) The `install-user` patch comes before the addition of `install` because it's more obviously a direct translation of the instructions in the README. The addition of `install` exposes and uses the `DESTDIR` variable so we can adjust the target installation path, which I then make use of to (re-)implement `install-user`. Finally the documentation is updated separate to either patch so if it's still contentious whether the patches should apply to master or not you can pick the others into `develop` while that discussion continues.

The targets still depend on `dist`, I haven't tackled the idea of making them instead depend on `compile`.

benjaoming commented on README.rst:
```diff
@@ -68,7 +68,17 @@ Build a fresh distribution package::
+defaults to ``DESTDIR=/usr/local``).
+
+Otherwise, on other branches, follow the steps below to manually install the
+distribution archive::
```
:100:

benjaoming reviewed (approved):
This is awesome! Thanks so much @amboar

mwilck commented:
@amboar ,

> To be honest I didn't expect the PR to drive so much discussion. 

That's how it goes when you make contributions to OSS projects, right? No matter how simple you think it would be, suddenly it gets a big issue :grin:   
I'm sorry that I had my part in that, I guess I was a bit overzealous.

Thanks a lot for your contribution.

amboar commented:
> @amboar ,
> 
> > To be honest I didn't expect the PR to drive so much discussion.
> 
> That's how it goes when you make contributions to OSS projects, right? No matter how simple you think it would be, suddenly it gets a big issue grin
> I'm sorry that I had my part in that, I guess I was a bit overzealous.

No, don't apologise! I'm just a drive-by contributor and have zero stake in ongoing maintenance, so I would have been fine with any response let alone possibly getting it merged. It just seemed like a simple enough change that I was just expecting "yeah, let's merge" or "nah, reasons, closed" 🙂 

> 
> Thanks a lot for your contribution.

No worries - thanks for taking the time to look at it and respond.

benjaoming commented on README.rst:
```diff
@@ -68,7 +68,17 @@ Build a fresh distribution package::
 
-Installing directly from file::
+
+*Shortcut on ``develop``*
```
```suggestion
**Shortcut on** `develop`
```

(the inline code didn't render well on GH)

benjaoming committed:
Update reST syntax

benjaoming commented on README.rst:
```diff
@@ -68,7 +68,17 @@ Build a fresh distribution package::
 
-Installing directly from file::
+
+**Shortcut on** `develop`
```
```suggestion
**Shortcut on** ``develop``
```

benjaoming committed:
Update reST syntax take 2

benjaoming commented:
This is a positive improvements that received some suggestions, but no real concerns, so it should be good to go.

(sorry about the extra commit, using Github Suggestions wasn't ideal for quick fingers)

DirkHoffmann commented:
Hi guys, 
This seems to have converged. I just "resolved" (I.e. collapsed) the three "conversations" for cleanup.
This PR now needed an assignee, and while every one of the reviewers would be fine, I have randomly choosen the first who answered. Anyone who is faster can change this again. :-) 
@amboar, just to give you some context, the four reviewers/managers who answered on this ticket are only in charge since March this year. So your proposal came in a period, where we still have to find our way with the github process, and it served as an example for that as well. Thank you very much for your input and fast feedback!
